src/onegov/search/indexer.py

```diff
@@ -306,8 +307,105 @@ def delete(self, task):
                     pass
 
 
-class TypeMapping:
+class PostgresIndexer(Indexer):
+    TEXT_SEARCH_COLUMN_NAME = 'fts_idx'
+
+    def __init__(self, mappings, queue, es_client, engine, session):
+        super().__init__(mappings, queue, es_client, hostname=None)
+
+        self.engine = engine
+        self.session = session
+
+        self.idx_language_mapping = {
+            'de': 'german',
+            'fr': 'french',
+            'it': 'italian',
+            'en': 'english',
+        }
+
+    def index(self, tasks) -> bool:
+        """ Update the 'fts_idx' column (full text search index) of the given
+        object(s)/task(s).
+
+        In case of a bunch of tasks we are assuming they are all from the
+        same schema and table in order to optimize the indexing process.
+
+        :param tasks: A list of tasks to index
+        :return: True if the indexing was successful, False otherwise
+        """
+        content = []
+
+        if not isinstance(tasks, list):
+            tasks = [tasks]
+
+        try:
+            for task in tasks:
+                language = (
+                    self.idx_language_mapping.get(task['language'], 'simple'))
+                data = {
+                    k: str(v)
+                    for k, v in task['properties'].items()
+                    if not k.startswith('es_')}
+                _id = str(task['id'])
+                content.append(
+                    {'language': language, 'data': data, '_id': _id})
+
+            schema = tasks[0]['schema']
+            tablename = tasks[0]['tablename']
+            id_key = tasks[0]['id_key']
+            table = sqlalchemy.table(
+                tablename,
+                (id_col := sqlalchemy.column(id_key)),
+                sqlalchemy.column(self.TEXT_SEARCH_COLUMN_NAME),
+                schema=schema  # type: ignore
+            )
+            tsvector_expr = sqlalchemy.text(
+                'to_tsvector(:language, :data)').bindparams(
+                sqlalchemy.bindparam('language', type_=sqlalchemy.String),
+                sqlalchemy.bindparam('data', type_=sqlalchemy.JSON)
+            )
+            stmt = (
+                sqlalchemy.update(table)
+                .where(id_col == sqlalchemy.bindparam('_id'))
+                .values({self.TEXT_SEARCH_COLUMN_NAME: tsvector_expr})
+            )
+            connection = self.engine.connect()
+            trans = connection.begin()
+            connection.execute(stmt, content)  # executemany supported with
+            # sqlalchemy >=1.4
+            trans.commit()
+        except Exception as ex:
+            index_log.error(f'Error \'{ex}\' indexing object id \'{_id}\' '
+                            f'in schema \'{schema}\' of table \'{tablename}\'')
+            return False
+
+        return True
+
+    def delete(self, task):
+        return True
 
+    def bulk_process(self):
```

Comment:
Should I index in batches of e.g. 1000  tasks?

For Meggen the number of documents/tasks are:
```
2024-03-07 13:31:51 - INFO: Indexing schema onegov_town6-meggen..               
2024-03-07 13:31:51 - INFO: Number of models to be indexed: 39                  
2024-03-07 13:32:14 - INFO: Indexing 3 tasks for directories                    
2024-03-07 13:32:14 - INFO: Indexing 334 tasks for pages                        
2024-03-07 13:32:14 - INFO: Indexing 7 tasks for external_links                 
2024-03-07 13:32:14 - INFO: Indexing 36 tasks for users                         
2024-03-07 13:32:14 - INFO: Indexing 4 tasks for filesets                       
2024-03-07 13:32:14 - INFO: Indexing 211 tasks for people                       
2024-03-07 13:32:14 - INFO: Indexing 3 tasks for resources                      
2024-03-07 13:32:14 - INFO: Indexing 167 tasks for directory_entries            
2024-03-07 13:32:14 - INFO: Indexing 315 tasks for events                       
2024-03-07 13:32:14 - INFO: Indexing 64 tasks for forms                         
2024-03-07 13:32:14 - INFO: Indexing 2969 tasks for tickets                     
2024-03-07 13:32:14 - INFO: Indexing 357 tasks for files                        
2024-03-07 13:32:14 - INFO: Done  
``

Reply:
Good question, not sure if it will be worth batching, but considering that on a complete re-index we will receive the tasks pre-sorted by table it might be more memory-efficient to use `itertools.groupby` and immediately call `index` on each group, rather than first collect all of the tasks into a dictionary.

So do something closer to the elasticsearch version, where you define a generator function, then use `itertools.groupby` to group the tasks returned by the generator and submit the group as a batch. If SQLAlchemy struggles with large batches we can always use `onegov.core.utils.batched` to split each group into smaller batches.